server/prompt_improvement_skill.py

```diff
@@ -0,0 +1,64 @@
+from pydantic import BaseModel, field_validator
+from adala.skills import Skill
+from typing import Any, Dict, List, Tuple
+from adala.skills.collection.text_generation import TextGenerationSkill
+
+
+class PromptImprovementSkillResponseModel(BaseModel):
+    
+    # hidden variable, used for validation
+    _input_variables: List[str]
+    reasoning: str
+    improved_user_prompt: str
+    # NOTE: not exposed in LSE yet, so default is always used. Should improve this as well when we expose it.
+    # improved_system_prompt: str
+
+
+    @field_validator("improved_user_prompt", mode="after")
+    def validate_used_variables(cls, value: str) -> str:
+
+        start_variable_idx = value.find("{")
+        end_variable_idx = value.rfind("}")
+        if start_variable_idx == -1 or end_variable_idx == -1 or start_variable_idx >= end_variable_idx:
+            raise ValueError("At least one input variable must be used in the prompt")
+
+        try:
+            value.format(**{var: "value" for var in cls._input_variables})
+        except KeyError as e:
+            raise ValueError(f"Invalid variable used in prompt: {e}. Valid variables are: {cls._input_variables}")
+
+        return value
+    
+
+def get_prompt_improvement_inputs(student_skill: Skill, input_variables: List[str], student_model: str) -> Dict[str, Any]:
+    return {
+        "model": student_model,
+        "task_name": student_skill.name,
+        "task_description": student_skill.description,
+        "input_variables": input_variables,
+        "current_system_prompt": student_skill.instructions,
+        "current_user_prompt": student_skill.input_template,
+        "response_json_schema": student_skill.response_model.model_json_schema(),
+    }
+
+
+def get_prompt_improvement_skill(input_variables: List[str]) -> TextGenerationSkill:
+
+    # setting this dynamically - used to validate the improved prompt
+    PromptImprovementSkillResponseModel._input_variables = input_variables
+    
+    prompt_improvement_skill = TextGenerationSkill(
+        name="prompt_improvement",
+        instructions="Improve the user prompt for the provided LLM model to complete the task using the provided input variables, with the provided user prompt as a starting point. Variables can be accessed in the user prompt using the format {variable_name} (only the variable values are used, not their names). Make sure your prompt produces output that will continue to conform to the provided json schema. Provide your reasoning for the changes you made to the prompt.",
+        input_template='''
```

Comment:
Let's move it to a separarate file, so we could keep refining it easily.
A couple of a prompt eng best practices that we could incorporate right away:

- split each variable in a well defined block, for example: `Model: {model}` --> `## Model\n{model}\n\n`
- add fewshot examples
- add `# Instructions:` or `# Follow the steps:` section with a numbered list of instructions
- add `# Things to avoid:` section where we patch the mistakes 

prompt engineering for refinement is out of the scope for sure, but moving `input_template` in a separate file and leaving `instructions` empty would allow us to easily iterate on that in the future updates

Reply:
I want to keep all the deps for the PromptImprovementSkill together, not sure putting just the input template in a separate file would be any cleaner.

But one thing I want to get your opinion on re prompt engineering is the trick of factoring out all the prompt contents that don't contain variables into the system prompt, and leaving the variables in the user prompt. This way we can easily use prompt caching on the system prompt, and it would also change the performance in some way - not sure if positive or negative, but I suspect positive, and I wanted to trial it here first, and if it works here bring it to LSE. wdyt?

Reply:
Updated the prompt. Keeping as separate system and user prompts for now, lmk if you have a strong opinion about that